aten/src/ATen/quantized/Quantizer.h

```diff
@@ -0,0 +1,226 @@
+};
+
+/**
+ * UniformQuantizer is the parent class for all uniform quantizers.
+ * These quantization scheme will map float value uniformly to
+ * the quantized value. For example, affine quantizer is
+ * the most commonly used scheme in this category.
+ */
+struct CAFFE2_API UniformQuantizer : public Quantizer {
+  UniformQuantizer() {}
+  UniformQuantizer(QScheme qscheme): Quantizer(qscheme) {}
+  virtual ~UniformQuantizer();
+};
+
+/**
+ * NonUniformQuantizer is the parent class for all non-uniform quantizers.
+ * These quantization scheme may map float value non-uniformly to the quantized
+ * value. K-means quantization is a representative example in this category.
+ */
+struct CAFFE2_API NonUniformQuantizer : public Quantizer {
+  NonUniformQuantizer() {}
+  NonUniformQuantizer(QScheme qscheme): Quantizer(qscheme) {}
+  virtual ~NonUniformQuantizer();
+
+};
+
+// There is also StochasticQuantizer which is uniform but not affine
+
+/**
+ * AffineQuantizer uses affine transformation to do quantization.
+ *
+ * For quantize:
+ * Y = clamp((X * scale + zero_point, min, max)
+ * For dequantize:
+ * X = (Y - zero_point) / scale
+ */
+struct CAFFE2_API AffineQuantizer : public UniformQuantizer {
+  AffineQuantizer() {}
+  AffineQuantizer(QScheme qscheme): UniformQuantizer(qscheme) {}
+  virtual ~AffineQuantizer();
+};
+
+/**
+ * SymmetricQuantizer is similar to AffineQuantizer except that it
+ * does not have zero_point
+ *
+ * For quantize:
+ * Y = clamp(X * scale, min, max)
+ * For dequantize:
+ * X = Y / scale
+ */
+struct CAFFE2_API SymmetricQuantizer : public UniformQuantizer {
+  SymmetricQuantizer() {}
+  SymmetricQuantizer(QScheme qscheme): UniformQuantizer(qscheme) {}
+  virtual ~SymmetricQuantizer();
+
+};
+
+/**
+ * PerTensorSymmetricQuantizer stores a single scale number which is
+ * used for quantizing all the values in the given Tensor
+ */
+struct CAFFE2_API PerTensorSymmetricQuantizer: public SymmetricQuantizer {
+  PerTensorSymmetricQuantizer() {}
+  PerTensorSymmetricQuantizer(float scale) : SymmetricQuantizer(kPerTensorSymmetric), scale_(scale) {}
+  virtual ~PerTensorSymmetricQuantizer();
+  float scale_{1.0};
+};
+
+/**
+ * PerChannelSymmetricQuantizer stores a vector of scale number and
+ * applys symmetric quantization using different scales on each channel.
+ *
+ * Also note that per channel quantization is mostly applied to output channels of
+ * weights since per-input channel of weight quantization or per-channel quantization
+ * for activations can't be efficiently supported in most of processors since
+ * it requires each multiplication result within a single dot-product
+ * to have a different scale.
+ */
+struct CAFFE2_API PerChannelSymmetricQuantizer: public SymmetricQuantizer {
+  PerChannelSymmetricQuantizer() {}
+  PerChannelSymmetricQuantizer(std::vector<float> scales, std::vector<int64_t> axis): SymmetricQuantizer(kPerChannelSymmetric), scales_(scales), axis_(axis) {
+    AT_ASSERT(axis_.size() == 1);
```

Comment:
AT_CHECK and add descriptive error message

Reply:
AT_CHECK or AT_ASSERTM? I saw a comment: 
// TODO: merge AT_CHECK with AT_ASSERTM. CHECK in fbcode means strict failure if
// not met.